### ISSUE TITLE
Ticket Redemption Service

### ISSUE THREAD

**Issue by kyriediculous** [livepeer/go-livepeer#1486]
# Ticket Redemption Service 

## Abstract

One desirable feature of a multi-orchestrator setup whereby each orchestrator uses a different keypair [1] is that not each orchestrator needs to be funded to execute transactions on the Ethereum network. 

One of these transactions are winning PM ticket redemptions. We can create a microservice that submits winning tickets on-chain on behalf of one or multiple orchestrator nodes.

While this service serves a clear purpose in a multi-O setup it could also be used in a single-O setup to seperate concern of keys. 

[1] https://github.com/livepeer/go-livepeer/pull/1476

### Repurpose `SenderMonitor`

The current `SenderMonitor` implementation keeps track of a sender's `maxfloat`. While a ticket redemption is pending on-chain the facevalue of the ticket will be substracted from the `maxFloat` until it is redeemed. These operations are exposed through the `SenderMonitor` API but these state updates could very well be local to the `SenderMonitor`.

In order to do so the current `SenderMonitor` implementation has to be the consumer of winning tickets instead of providing a channel to pass-through tickets from `recipient` to `ticketqueue` and vice versa. The `SenderMonitor` already has access to the `Broker` interface so this is easily achievable and reduces the API surface even further. 

**API**

```
type SenderMonitor interface {
	// Start initiates the helper goroutines for the monitor
	Start()

	// Stop signals the monitor to exit gracefully
	Stop()

	// QueueTicket adds a ticket to the queue for a remote sender
	QueueTicket(addr ethcommon.Address, ticket *SignedTicket) error

	// MaxFloat returns a remote sender's max float
	MaxFloat(addr ethcommon.Address) (*big.Int, error)

	// ValidateSender checks whether a sender's unlock period ends the round after the next round
	ValidateSender(addr ethcommon.Address) error
}
```

### Ticket Redemption Service

This service will be responsible for two things: 

1. Redeeming winning tickets received from nodes within a cluster of orchestrators. Tickets will be added to a **queue** from which tickets will be popped when there are no pending ticket redemptions on-chain. 

2. Send `maxfloat` updates for ticket senders to all orchestrators in the cluster.

The server will keep track of all client streams to send messages to and store it as state. When a new ticket redemption comes in from one of the clients the server will loop over all streams and send a message. 

![](https://i.imgur.com/2B1YyqH.png)

**gRPC Server**
Create a **Ticket Redemption Service**. This is a gRPC microservice that can be started when running a multi-O cluster of which the client implements the `TicketRedeemer` interface.
 
The server will largely use the `senderMonitor` implementation for its business logic:
- Watching on-chain events pertaining to `Reserve` 
- Keeping track of max float
- Add tickets to the `ticketqueue`
- Consume winning tickets through the `Broker` interface. 

The **Ticket Redemption Service** will be implemented as a **gRPC** service:
- Unary RPC call to submit tickets 
- Server-side streaming for `maxfloat` updates 

Clients (orchestrators) will subscribe to `MaxFloat` and listen for new messages received on the stream. They can then submit winning tickets by calling the `QueueTicket` RPC method. 

When the redemption service receives a winning ticket, it will subtract the ticket facevalue from the sender's  `maxfloat`. An update with the new `maxfloat` value will be broadcasted to all orchestrators. When the winning ticket is redeemed successfully on-chain, the redeemed amount will be added back to the `maxFloat` and again a message will be sent to the clients with the updated value. 

**gRPC Client**

The clients will implement the `SenderMonitor` interface, while `QueueTicket` will be a simple pass through, the client does keep track of the latest received `maxFloat` value for a sender and provides a memory cache to other operations running on the orchestrator node (such as `TicketParams`  generation). 

#### Proto definitions

```proto
service TicketRedeemer {
    rpc QueueTicket(Ticket) returns (google.protobuf.Empty) {}
    rpc MonitorMaxFloat(google.protobuf.Empty) returns (stream MaxFloatUpdate) {}
}

message Ticket {
    bytes recipient = 1;
    bytes sender = 2;
    bytes face_value = 3;
    bytes win_prob = 4;
    uint32 sender_nonce = 5;
    bytes recipient_rand_hash = 6;
    int64 creation_round = 7;
    bytes creation_round_block_hash = 8;
    bytes sig = 9;
    bytes recipient_rand = 10;
}

message MaxFloatUpdate {
        bytes sender = 3;
        bytes max_float= 4;
}

```

* We could use a single bi-directional stream but that would require `oneof` nesting in messages because an orchestrator is likely to need to subscribe to the `Redemption` pubsub before it send a winning ticket. Having separate streams and unnested message definitions allows use more flexibility in terms of backwards incompatible changes going forward. 

### Engineering Tasks 

1. Create a `TicketRedeemer` interface

2. Implement current ticket redemption logic as `TicketRedeemer` interface

3. Create Ticket Redemption Service RPC with an API that implements `TicketRedeemer` interface 

4. Allow running a ticket redemption service by providing startup flags

5. (Allow ticketservice to be a standalone service (separate binary) )
### Authorization

The current architecture is very well suited for TLS certificates (either self-signed or user provided). Designing a proper TLS authorization scheme however is out of scope for this project. Therefore authorization can be added in a later iteration. 

## Known Unknowns

It's unclear what a full multi-o setup will look like at this point. 

## Alternatives

Other networking layer alternatives for doing HTTP push include:

- (long) polling
- websockets
- webhooks

There's some clear benefits to using gRPC over HTTP(1) REST or JSON-RPC as it supports HTTP2 out of the box. We already use gRPC for node-to-node communications so it makes sense to use it for this service as well. 

Don't use bi-directional streams and divide the **max float** by the number of nodes the orchestrator operates. This has a negative impact on the number of orchestrators can be ran in a cluster. 

## Additional Context

The gRPC client-server structure could also be used to reduce the number of calls to remote ethereum nodes (eg blockwatcher) whereby a service listens for events on behalf of all O's and propagates them across the network through client side streams.

**Comment by yondonfu** [comment]
> Tickets will be added to a queue from which tickets will be popped when there are no pending ticket redemptions on-chain.

Could we use the existing queuing from `pm/queue.go` i.e one queue per sender and redeeming tickets serially when their ticket params have expired?

We'll also want the service to store the ticket (i.e. write to DB) so that if it goes down it can still redeem them after restarting. We can probably start by just storing the ticket upon receiving it on the service similar to what we do in `pm.ReceiveTicket()` [1]. After completing a MVP version of the ticket redemption service, implementing #942 could be useful so that Os with or without an external ticket redemption service will be able to redeem stored tickets after a restart.

> Send messages to all orchestrators in the cluster about a ticket redemption lifecycle (received, pending, mined)

I think Os would only need to know about pending redemptions (O needs to subtract the ticket face value from the sender's current max float) and mined redemptions (O needs to add the value redeemed to the sender's current max float).

Worth noting that `pm.Recipient` currently adds the ticket face value to the sender's max float after a redemption is mined, but we actually want to add the value redeemed (since the value redeemed can be less than the ticket face value if the sender's reserve cannot cover the full ticket face value) as mentioned in this [TODO](https://github.com/livepeer/go-livepeer/blob/0ee99d8905d461d734b638b0f5b533caa3e2a016/pm/recipient.go#L350). The `Redemption` message could include the value redeemed instead of the ticket face value.

> (Allow ticketservice to be a standalone service (separate binary) )

WDYT about excluding this from the scope of this project, focusing exclusively running the service on an O node initially (since we can run the reward service on that O node as well) and then considering a standalone binary separately?

**Comment by kyriediculous** [comment]
> Could we use the existing queuing from pm/queue.go i.e one queue per sender and redeeming tickets serially when their ticket params have expired?

No the ticket redemption service will have no idea of the `maxfloat`. 
The problem is that we have multiple concurrent services submitting tickets which could rapidly decrease the maxfloat to a point whereby it has not enough reserve to cover new tickets _when calculating ticket params_ until the outstanding ones have been redeemed.  So there's two solutions 

1. On the redemption services redeem concurrent received tickets serially. 
2. Make sure max float always remains above 1x  (or a multiplier) of cfg.EV 

Given the likelihood of concurrent winning tickets is rather low (with proper configuration) I feel like just having the redemption service redeem winning tickets serially is the simplest solution with the least room for error or unexpected behaviour. 

> We'll also want the service to store the ticket (i.e. write to DB) so that if it goes down it can still redeem them after restarting.

In my philosophy the redemption service would be very light weight. The separate clients will be responsible for handling failures and storing winning tickets.

When the redemption failure for example returns an error tickets will be added to the local `ticketqueues` from their origin orchestrator node. 

Nodes can be responsible for storing tickets in their own DB and submitting them to their `ticketqueues` on startup. These queues will then submit the tickets to the redemption service when the necessary conditions are met. 

> I think Os would only need to know about pending redemptions (O needs to subtract the ticket face value from the sender's current max float) and mined redemptions (O needs to add the value redeemed to the sender's current max float).

In the origin implementation we sub the facevalue from the maxfloat before submitting the transaction. But since the intention is to have it subtracted while the tx is pending we can notify the nodes after submission as well, good point 👍 

> The Redemption message could include the value redeemed instead of the ticket face value.

Thanks for the heads up. Since this would require changes to the current implementation as well what do you think about tackling this as a separate issue after we finish the ticket redemption service?  I can add it as a last implementation task. 

> WDYT about excluding this from the scope of this project

Yeah definitely, I also intend for these bullet points to be separate issues to work on.

**Comment by j0sh** [comment]
>     rpc RedeemWinningTickets(stream Ticket) returns (stream google.protobuf.Empty) {}

Does this particular call need to be a bidirectional stream? Would this work as a plain request-reply RPC call? The return value could be some sort of acknowledgement that the ticket was queued, eg

```
rpc RedeemWinningTickets(Ticket) returns (TicketAck);
```

> Send messages to all orchestrators in the cluster about a ticket redemption lifecycle (received, pending, mined) to orchestrators to independently update their max float value for the sender.
>
> No the ticket redemption service will have no idea of the maxfloat.

I don't quite understand the reasoning for making the redemption service unaware of the max float (or rather, the current float). It seems that making the service float-aware could make the architecture simpler. Ticket state updates would not have to be propagated to all nodes (with all the synchronization that entails), only the absolute value of the current float for a given B that the O is interested in. The subscription could look something like this:

```
rpc MonitorFloat(BroadcasterAddress) returns (stream FloatValue);
```

> When the redemption failure for example returns an error tickets will be added to the local ticketqueues from their origin orchestrator node.

Do you mean the redemption service would not be responsible for retrying, but rather the origin orchestrator would be responsible for storing the ticket, monitoring redemption status, and re-submitting to the redemption service as necessary? I'm curious what are the upsides to that pattern, because I see a few downsides:

Orchestrators lose ephemerality if they're responsible for tracking the state of the ticket and re-trying. Ideally, ticket submission could be a fire-and-forget affair, and the O could be cleaned up immediately afterwards without consequence. Operationally, this seems simpler, and enables more elasticity among the set of nodes.

Keeping state in sync across both the O node and the redemption service also seems more error-prone. For example, suppose either the O node or the redemption service goes down in the middle of a state transition. Keeping any recovery logic isolated to the redemption service alone feels like it'd be easier to reason about. This feeds back into the earlier note about simply subscribing and transmitting absolute numbers for the current float.

> WDYT about excluding this from the scope of this project, focusing exclusively running the service on an O node initially (since we can run the reward service on that O node as well) 

What we could do is offer an option for running the go-livepeer binary in "standalone redemption" mode. I think this is mostly a matter of disabling the transcoder server at startup:

https://github.com/livepeer/go-livepeer/blob/716119b0e6150d04ad3c6146aea4f9ab3114a820/cmd/livepeer/livepeer.go#L774-L777

I think things should naturally fall together in a composable manner. Eg:
`-orchestrator -redeemerAddr` runs an orchestrator connecting to the given redeemer. Omitting the `-orchestrator` flag, supplying only a `-redeemer` flag could run things in standalone redeemer mode. `-orchestrator` alone would keep the current behavior of an all-in-one orchestrator - I believe it's important to maintain CLI flag compatibility wherever possible, to ease the transition for users during upgrades.

Note that I suspect we may also want this service to do other blockchain-related things, such as invoking `reward` so disabling (or enabling) the redemption service could also toggle a few other blockchain services. (Minor point of naming, but should we call it something other than the "redemption service" if the intent is to eventually provide a more generic entry point for blockchain monitoring and transaction services?)

**Comment by kyriediculous** [comment]
> Does this particular call need to be a bidirectional stream? Would this work as a plain request-reply RPC call? The return value could be some sort of acknowledgement that the ticket was queued, eg

I'd be okay turning this into a unary RPC call

> It seems that making the service float-aware could make the architecture simpler. Ticket state updates would not have to be propagated to all nodes (with all the synchronization that entails), only the absolute value of the current float for a given B that the O is interested in. 

I initially considered this. The reason why I didn't choose to go this route for now because I see fitting that in a later iteration of this service. If the ticket redemption service were to be the central source for `current float` it would have to watch for on-chain events as well. So I feel it doesn't make the architecture simpler _at this point_. 

The benefit of the current architecture is that this is a very lightweight service that has a single job and that gives us maximum flexibility going forward once we have a better sight on a final design for a proper multi-O setup. It fits really well into the current workflow in that it doesn't change much. 

E.g.  Currently orchestrators watch for on-chain events but to reduce the number of on-chain requests to remote nodes (especially when using a hosted service like infura) we could introduce a central service that watches for events and propagates these to the orchestrators or other central services that keep state, such as potentially the ticket redemption service for `RedeemWinningTicket` and `FundReserve` events to keep track of the current float. 

> Do you mean the redemption service would not be responsible for retrying, but rather the origin orchestrator would be responsible for storing the ticket, monitoring redemption status, and re-submitting to the redemption service as necessary?

> Orchestrators lose ephemerality if they're responsible for tracking the state of the ticket and re-trying.

Correct this is how it already works, except that tickets would be submitted to an external service rather than locally so it introduces minimum changes to the current workflow. I acknowledge though that there are benefits to having the ticket redemption service be responsible for this. But again we don't really have a full design out for a multi-O setup. 

Once we have a better idea on that front I think we can implement that. I feel like we first need to think or at least settle on what a mutli-O setup would look like. 

E.g. In this context will the ticket redemption service have its own DB or will one orchestrator in the cluster be responsible for running the ticket redemption service and granting it access to its database and considering this as a "master" orchestrator that is non-emphemeral and do we really expect orchestrator nodes to be  _that_ ephemeral, I see that happening more in the context of remote transcoders (eg burst capacity). 

____

tl;dr designed for simplicity and minimal changes to current code which gives us more leeway to design a multi-O setup and move to a richer ticket redemption service in the future.

**Comment by iameli** [comment]
I like the overall design, and I like the movement toward externalizing a "blockchain service" more generally. I also agree with @j0sh that from an orchestrator's perspective, ticket redemption should be "fire and forget" with all retry logic happening on the redeemer's side after the point that a `RedemptionsRequest` is successfully called.

> I feel like we first need to think or at least settle on what a mutli-O setup would look like.

Agreed, but let's do it now — @yondonfu already has a version of this running.

> E.g. In this context will the ticket redemption service have its own DB or will one orchestrator in the cluster be responsible for running the ticket redemption service and granting it access to its database and considering this as a "master" orchestrator that is non-emphemeral

I do think it will be very important for the ticket redemption service to have its own DB, so as to maximize the durability of winning tickets. We also need to consider, if we're really talking about arbitrary capacity sitting behind a single O address, what it means to operate such a service with global availability. Certainly that full design exceeds the scope of this proposal, but that's the direction I see us heading. Our check cashing service needs to be careful with its data.

> and do we really expect orchestrator nodes to be that ephemeral, I see that happening more in the context of remote transcoders (eg burst capacity).

We got here by working on this design _instead_ of doing more work and testing on Multi-T. I think that's a good prioritization, but it does mean that "burst capacity" is exactly one of the use cases we should account for in this design. So yes, I fully expect these ephemeral orchestrators to go up and down a lot, scale with traffic, and roll over as we deploy new versions of go-livepeer. In my ideal world, the _only_ time that terminating an orchestrator would be a problem is if its received a winning ticket but hasn't submitted it to the redemption service yet. Are there other issues I'm not considering?

A thought: I'm an orchestrator. Everything got set up properly, but now I've been trying to connect to my `-redeemerAddr` for (IDK, perhaps) 60 seconds without success. What do I do?

1. Crash. If I can't cash out tickets or query pertinent float balances, I shouldn't keep running.

2. Keep running, but turn down any new work. Keep trying to hit the `-redeemerAddr` forever. Start accepting work again once I reconnect.

3. Keep running and keep transcoding. This maximizes my service availability, but runs the risk that winning tickets would be lost if my pod gets deleted or whatever.

**Comment by j0sh** [comment]
>  If the ticket redemption service were to be the central source for current float it would have to watch for on-chain events as well

Don't we already do this? Because we're sharing the same codebase, could the on-chain monitoring also be initialized for the redemption service? (And maybe correspondingly disabled on the orchestrator nodes that aren't doing their own redemptions.)

> It fits really well into the current workflow in that it doesn't change much.

From a quick glance through the code, I think we're very close to being able to substitute out the relevant parts without major workflow changes. The `pm.Recipient.RedeemWinningTicket` function call can be substituted for a RPC call as appropriate - fire and forget. AFAICS, other than the MaxFloat, all the current SenderMonitor usage revolves around managing ticket submissions. For an orchestrator that outsources ticket submission, the "sender monitor" given to the constructor of this "recipient RPC client" can consist of a minimal interface that only defines a MaxFloat for a given B, and does a RPC subscription under the hood to update the local value as needed.

Note that we should be able to keep much of the current redemption code as-is, and have it run in "redeemer mode" - whether the redeemer is standalone, or part of an orchestrator. The orchestrator node gets initialized with a "RPC recipient" or a "classic recipient". A standalone redeemer gets a "classic recipient" that's fed via RPC calls, rather than from within `core/orchestrator.go`.

We *might* have to add a few steps here and there - not sure if we should also have explicit subscribe / unsubscribe actions somewhere, eg as the session begins and ends. And maybe some scaffolding around broadcasting float updates, as well (perhaps by another `SenderMonitor` implementation that wraps the current `senderMonitor` with RPC handling).

> tl;dr designed for simplicity and minimal changes to current code which gives us more leeway to design a multi-O setup and move to a richer ticket redemption service in the future.

As @iameli mentions, this is our opportunity to figure out what a multi-O service should look like, and design towards that. This is one of those cases where it's a good tradeoff to put in a bit more upfront effort to avoid the code / design churn that would otherwise be inevitable.

> I do think it will be very important for the ticket redemption service to have its own DB, so as to maximize the durability of winning tickets. 

Yep. Although note that we don't persist winning tickets to DB right now, so we can probably continue to live without it as a first pass. But it should go in pretty soon. Following up on a couple of the other DB-related threads:

> In this context will the ticket redemption service have its own DB or will one orchestrator in the cluster be responsible for running the ticket redemption service and granting it access to its database and considering this as a "master" orchestrator that is non-emphemeral

Since we're sharing the same codebase - we can put the redemption info into the DB we already have. There are two cases I can think of here:
* Standalone redemption service. Has its own DB (really, the same DB we already have)
* Orchestrator that does its own redemptions. Already has its own DB.

I'm not quite sure what's meant by the suggested alternative of "granting access" though... is it something like this?
* Orchestrator does *not* do its own redemptions. But there's also a separate redemption process, using the same working directory, which reads and writes from the orchestrator's DB.

The above does not seem like something we should encourage, if only because of sqlite's single-writer restriction.

Should we also allow orchestrator nodes that do their own redemptions to accept redemptions on behalf of other nodes? Dunno - will leave that up to you. Probably just a matter of toggling the RPC endpoint, and it'd be trivial to populate a DB-backed queue from multiple sources. As a matter of practice though, I'd avoid mixing hot keys with an orchestrator node.

>> After completing a MVP version of the ticket redemption service, implementing #942 could be useful 
>>
> Nodes can be responsible for storing tickets in their own DB and submitting them to their ticketqueues on startup.

Given that DB updates will be required in the current queuing code to update the state of a given ticket, I'd second @yondonfu 's suggestion to implement #942 . With the queue implemented in terms of the DB (rather than having the DB simply feed the current queue), I suspect we'll be able to remove a lot of complexity from the current queue, while gaining robustness.

> I've been trying to connect to my -redeemerAddr for (IDK, perhaps) 60 seconds without success. What do I do?
>
> 2. Keep running, but turn down any new work. Keep trying to hit the -redeemerAddr forever. Start accepting work again once I reconnect.

I'd maybe modify Option 2 and optimistically continue to perform work for existing sessions, but reject new sessions. Presumably the float will be available locally for existing sessions, and the float for a given B is not likely to change dramatically over the span of a few minutes. Since the redeemer service is unavailable, the O can't determine the float available for new sessions, so those need to be rejected.

Another option for degraded continuity of service: if the redeemer service is unreachable, we could fall back to doing blockchain monitoring of the float from the local node. Once the connection to the redeemer service is re-established, the tickets will be submitted and float monitoring should reset the node's view of a given float. Likely out of scope for a while though.

For now, though, the redemption service going down - especially if it appears down for multiple orchestrator nodes - is probably cause for waking up the on-call person ;)

**Comment by kyriediculous** [comment]
Thanks for the feedback , after a discussion with @yondonfu earlier this week the go-ahead plan is as follows: 

1. Create a DB backed ticket retry queue and remove the `TicketStore` as a dependency for the recipient.  

WIP: https://github.com/livepeer/go-livepeer/compare/nv/db-ticketq

2. Rework the current `SenderMonitor` API to expose only `Start()`,  `Stop()` , `QueueTicket()` and `MaxFloat()`. Keep maxFloat updates internal to the `senderMonitor` implementation (whether that's the current local setup or the ticket redemption service).  The `SenderMonitor` will now also become the ticket consumer instead of `recipient`

WIP (minus DB queue): https://github.com/livepeer/go-livepeer/compare/nv/rework-sm

_`recipient` is now a fire and forget module_ 

3. The `SenderMonitor` implementation , including a DB backed ticketqueue can now be re-used for the server implementation of the ticket redemption service. A local setup and multi-o setup now share this implementation 

4. Write a second `SenderMonitor` implementation that is used in a multi-O setup which pubsubs for updated maxfloat values instead of listening for on-chain events and doing local updates for pending tickets.

5. Connect these services with the user interface by adding flags

I believe these action points address the feedback and concerns from the discussion. 

I will update the spec accordingly shortly.
